x-pack/test/plugin_functional/plugins/resolver_test/public/applications/resolver_test/index.tsx

```diff
@@ -30,34 +34,51 @@ export function renderApp(
   };
 }
 
-const AppRoot = React.memo(
-  ({ embeddable: embeddablePromise }: { embeddable: Promise<IEmbeddable | undefined> }) => {
-    const [embeddable, setEmbeddable] = React.useState<IEmbeddable | undefined>(undefined);
-    const [renderTarget, setRenderTarget] = React.useState<HTMLDivElement | null>(null);
-
-    useEffect(() => {
-      let cleanUp;
-      Promise.race([
-        new Promise<never>((_resolve, reject) => {
-          cleanUp = reject;
-        }),
-        embeddablePromise,
-      ]).then(value => {
-        setEmbeddable(value);
-      });
-
-      return cleanUp;
-    }, [embeddablePromise]);
-
-    useEffect(() => {
-      if (embeddable && renderTarget) {
-        embeddable.render(renderTarget);
-        return () => {
-          embeddable.destroy();
-        };
-      }
-    }, [embeddable, renderTarget]);
-
-    return <div data-test-subj="resolverEmbeddableContainer" ref={setRenderTarget} />;
-  }
-);
+const AppRoot = styled(
+  React.memo(
+    ({
+      embeddable: embeddablePromise,
+      className,
+    }: {
+      embeddable: Promise<IEmbeddable | undefined>;
+      className?: string;
+    }) => {
+      const [embeddable, setEmbeddable] = React.useState<IEmbeddable | undefined>(undefined);
+      const [renderTarget, setRenderTarget] = React.useState<HTMLDivElement | null>(null);
+
+      useEffect(() => {
+        let cleanUp;
```

Comment:
:grey_question:  If you're using cleanUp for something here, it is really not clear what. I see it gets the reject side of the Promise below, but I don't understand where or how it's used at all. Can you explain?

Reply:
I added comments to explain but i'll reiterate here. The 'effect' is responsible for waiting for a promise to resolve and then setting component 'state' from that resolved value.

There is complexity because the promise comes in as a React prop. The prop could change at any point. When the prop changes, we call `Promise.race`. If the value is resolved, we set the value in state. If the prop changes, we cancel the race by rejecting.